CIPs/0017.md

```diff
@@ -0,0 +1,69 @@
+
+# CIP [0017]: Integrate Plumo Proofs into the lightest sync protocol
+
+- Date: 2020-07-27
+- Author: @lucasege
+- Status: WIP
+
+## Overview
+
+In order to continue supporting the dream of a seamless mobile sync experience, the lightest sync protocol must be extended to support PLUMO Zk-SNARKs which can quickly verify transitions from epoch X to epoch X + Y, where Y could be on the order of 10s-100s of epochs.
+
+The prover/verifier for plumo proofs have already been implemented, and the exposed go interface is available [here](https://github.com/celo-org/celo-bls-go)
+
+## Goals
+
+- Support adding/storing proofs on full nodes
+- Gossip these proofs to propogate them through the network
+- Extend the `les` protocol such that the server will provide these proofs in place of headers (when appropriate), and the client will consume/verify proofs accordingly.
+
+## Proposed Solution
+
+We propose to add two new RPCs: 
+`AddProof(proof []byte, proofId uint)`
+and
+`Proofs() ([][]byte)`. When an eventual "proving service" generates a new proof it will publish this proof to a full node that it is running, which will then propogate the proof throughout the network.
+
+This `AddProof` function will store the proof in a new database `plumoproofdata`, keyed by a unique proof ID.
+
+To propogate these proofs throughout the network, we also propose to extend the `eth` protocol with two new message types: 
+```
+PlumoProofMsg - Sends plumo proof(s) either in response to a `GetPlumoProofMsg` or after a successful RPC call to `AddProof`
+GetPlumoProofMsg - Request plumo proof(s) from a peer
+```
+`GetPlumoProofMsg` will be implemented by having the requester send a set of proof IDs that they hold, and ask for any they are missing. This may be optimized using a bloom filter as well.
+
+The `les` protocol must then also be extended via at least two messages:
```

Comment:
Implementation note: `les/4` is pending on `master` to include gateway fee changes that are not fully complete. We will want to provide MVP support for those (which may even be a stub) or figure out another way to handle this if we want to have these modification as part of `les/4` (Another option is to move gateway fee messages to `les/5` as well if there is no time to get it fully ready)

Reply:
Hm good point, thanks for pointing this out. I'm inclined to go the second route (plumo = les/4, gateway fees to les/5), let me know if you agree or if this changes anything else @nategraf